pkg/simulator/objects/apply.go

```diff
@@ -0,0 +1,385 @@
+
+	// DiscoveryClient queries API server about the resources
+	dc, err := discovery.NewDiscoveryClientForConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+	mapper := restmapper.NewDeferredDiscoveryRESTMapper(memory.NewMemCacheClient(dc))
+
+	return mapper.RESTMapping(gvk.GroupKind(), gvk.Version)
+}
+
+//verifyObj is a helper method used to verify objects to make it easier to test
+func verifyObject(obj *unstructured.Unstructured, fn func(interface{}) (bool, error), keys ...string) (ok bool, err error) {
+	tmpObj, ok, err := unstructured.NestedFieldCopy(obj.Object, keys...)
+	// if not found or no key present
+	if err != nil || !ok {
+		return ok, err
+	}
+
+	if fn != nil {
+		return fn(tmpObj)
+	}
+
+	return ok, err
+}
+
+// cleanupObjects will clean up all "null" strings that appear in
+// support bundles.
+func cleanupObjects(obj map[string]interface{}) error {
+	// key: null is a valid value in prometheusrules, hence that is ignored from this cleanup
+	for key, value := range obj {
+		if v, ok := value.(string); ok && v == "null" && key != "key" {
+			delete(obj, key)
+		}
+
+		if _, ok := value.([]string); ok {
+			continue
+		}
+
+		if key == "resourceVersion" {
+			delete(obj, key)
+		}
+
+		if valArr, ok := value.([]interface{}); ok {
+			var newArr []interface{}
+			for _, v := range valArr {
+				newMap, innerOK := v.(map[string]interface{})
+				if innerOK {
+					err := cleanupObjects(newMap)
+					if err != nil {
+						return err
+					}
+					newArr = append(newArr, newMap)
+				}
+
+			}
+			if len(newArr) != 0 {
+				obj[key] = newArr
+			}
+		}
+
+		if valMap, ok := value.(map[string]interface{}); ok {
+			err := cleanupObjects(valMap)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// FetchObject will use the dynamic client to fetch runtime.Object from apiserver.
+func (o *ObjectManager) FetchObject(obj runtime.Object) (*unstructured.Unstructured, error) {
+	var dr dynamic.ResourceInterface
+	unstructObj, err := wranglerunstructured.ToUnstructured(obj)
+	if err != nil {
+		return nil, err
+	}
+	restMapping, err := findGVR(unstructObj.GroupVersionKind(), o.config)
+	if err != nil {
+		return nil, err
+	}
+	if restMapping.Scope.Name() == meta.RESTScopeNameNamespace {
+		dr = o.dc.Resource(restMapping.Resource).Namespace(unstructObj.GetNamespace())
+	} else {
+		dr = o.dc.Resource(restMapping.Resource)
+	}
+
+	return dr.Get(o.ctx, unstructObj.GetName(), metav1.GetOptions{})
+}
+
+// WaitForNamespaces ensures apiserver is ready and namespaces can be listed before it times out
+func (o *ObjectManager) WaitForNamespaces(timeout time.Duration) error {
+	now := time.Now()
+	for currtime := now; currtime.Before(now.Add(timeout)); {
+		// ignore errors
+		ns, _ := o.kc.CoreV1().Namespaces().List(o.ctx, metav1.ListOptions{})
+		if ns != nil && len(ns.Items) != 0 {
+			return nil
+		}
+		time.Sleep(5 * time.Second)
+	}
+
+	return fmt.Errorf("timed out waiting for apiserver to be ready")
+}
```

Comment:
Perhaps use `Poll`? https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait#Poll

Reply:
@weihanglo thanks for the suggestion. I have changed the logic to implement `wait.Poll`